cql3/restrictions/statement_restrictions.cc

```diff
@@ -1364,6 +1364,43 @@ void statement_restrictions::prepare_indexed(const schema& idx_tbl_schema, bool
         const auto pos = _schema->position(*col) + 1;
         (*_idx_tbl_ck_prefix)[pos] = replace_column_def(e, &idx_tbl_schema.clustering_column_at(pos));
     }
+
+    add_clustering_restrictions_to_idx_ck_prefix(idx_tbl_schema);
+
+    auto token_column = &idx_tbl_schema.clustering_column_at(0);
+    (*_idx_tbl_ck_prefix)[0] = binary_operator(
+            column_value(token_column),
+            oper_t::EQ,
+            // TODO: This should be a unique marker whose value we set at execution time.  There is currently no
+            // handy mechanism for doing that in query_options.
+            ::make_shared<constants::value>(raw_value::make_unset_value()));
+}
+
+void statement_restrictions::prepare_indexed_local(const schema& idx_tbl_schema) {
+    if (!_partition_range_is_simple) {
+        return;
+    }
+
+    // Local index clustering key is (indexed column, base clustering key)
+    _idx_tbl_ck_prefix = std::vector<expr::expression>();
+    _idx_tbl_ck_prefix->reserve(1 + _clustering_prefix_restrictions.size());
+
+    const column_definition& indexed_column = idx_tbl_schema.column_at(column_kind::clustering_key, 0);
+    const column_definition& indexed_column_base_schema = *_schema->get_column_definition(indexed_column.name());
+
+    // Find index column restrictions in the WHERE clause
+    std::vector<expr::expression> idx_col_restrictions = extract_column_restrictions(*_where, indexed_column_base_schema);
+    expr::expression idx_col_restriction_expr = expr::expression(expr::conjunction{std::move(idx_col_restrictions)});
+
+    // Translate the restriction to use column from the index schema and add it
+    expr::expression replaced_idx_restriction = replace_column_def(idx_col_restriction_expr, &indexed_column);
+    _idx_tbl_ck_prefix->push_back(replaced_idx_restriction);
+
+    // Add restrictions for the clustering key
+    add_clustering_restrictions_to_idx_ck_prefix(idx_tbl_schema);
```

Comment:
How does that work for a tuple? You replaced the column def you're looking for, but you didn't replace the other column defs. Those other column defs refer to the base table schema, not the index.

Reply:
(and might not even exist in the index)

Reply:
You're right, currently `multi-column` restrictions seem to be broken on indexes. 
I did the same thing that was done in the global index, but it looks like it doesn't handle `multi-column` restrictions at all.

I made a test:
```cql
DROP KEYSPACE IF EXISTS ks;
CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
CREATE TABLE IF NOT EXISTS ks.t (p int, c1 int, c2 int, c3 int, r1 int, r2 int, primary key (p, c1, c2, c3));
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 0, 0, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 1, 0, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 2, 2, 0, 0, 0);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 0, 1, 0, 0, 1);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 1, 1, 0, 0, 1);
INSERT INTO ks.t (p, c1, c2, c3, r1, r2) VALUES (0, 3, 3, 0, 0, 1);

-- Uncomment for a global index:
CREATE INDEX IF NOT EXISTS ON ks.t(r2);

-- Uncomment for a local index:
-- CREATE INDEX IF NOT EXISTS ON ks.t((p), r2);
SELECT * FROM ks.t WHERE p = 0 and (c1, c2) < (1, 0) and r2 = 0;
```
Prints out:
```cql
 p | c1 | c2 | c3 | r1 | r2
---+----+----+----+----+----
 0 |  0 |  0 |  0 |  0 |  0
 0 |  1 |  0 |  0 |  0 |  0
 0 |  2 |  2 |  0 |  0 |  0
```
(Sometimes the index doesn't build in time and the rows are empty)

I will open an issue for this

Reply:
I opened #9085

Reply:
Alright it seems that when there are `multi-column` restrictions we should not use an index at all, so we can assume that there are no `multi-column` restrictions in those expressions.

Reply:
We can't assume anything. We can have one restriction used for indexing, and another for filtering.

`WHERE c1 = 3 AND (c1, c2) = (3, 4)`. The first relation would be to hook to the index, the second is just a filter.

Reply:
But we agreed previously that we won't be using the index in such queries.

Reply:
Sure, we shouldn't be using the index, but the code shouldn't blow up. I think this means that extract_column_restrictions should not return multi-column restrictions.

Reply:
> extract_column_restrictions should not return multi-column restrictions

After more thinking, I now agree with this.  If we do use an index (something that currently happens and may continue to happen*), a multi-column restriction isn't helpful in determining the index-table clustering prefix.

*when a regular-column index combines with multi-column restrictions, we end up using the index; I tried changing that but concluded it's easier to keep it that way for now.